wgpu-core/src/resource.rs

```diff
@@ -1,29 +1,131 @@
 use crate::{
-    device::{DeviceError, HostMap, MissingDownlevelFlags, MissingFeatures},
+    device::{Device, DeviceError, HostMap, MissingDownlevelFlags, MissingFeatures},
     global::Global,
     hal_api::HalApi,
-    hub::Token,
-    id::{AdapterId, DeviceId, SurfaceId, TextureId, Valid},
+    id::{
+        AdapterId, BufferId, DeviceId, QuerySetId, SamplerId, StagingBufferId, SurfaceId,
+        TextureId, TextureViewId, TypedId, Valid,
+    },
     identity::GlobalIdentityHandlerFactory,
     init_tracker::{BufferInitTracker, TextureInitTracker},
     track::TextureSelector,
     validation::MissingBufferUsageError,
-    Label, LifeGuard, RefCount, Stored,
+    Label, SubmissionIndex,
 };
 
+use parking_lot::{Mutex, RwLock};
 use smallvec::SmallVec;
 use thiserror::Error;
 
-use std::{borrow::Borrow, ops::Range, ptr::NonNull};
+use std::{
+    borrow::Borrow,
+    fmt::Debug,
+    ops::Range,
+    ptr::NonNull,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+/// Information about the wgpu-core resource.
+///
+/// Each type representing a `wgpu-core` resource, like [`Device`],
+/// [`Buffer`], etc., contains a `ResourceInfo` which contains
+/// its latest submission index and label.
+///
+/// A resource may need to be retained for any of several reasons:
+/// and any lifetime logic will be handled by Arc<Resource> refcount
+///
+/// - The user may hold a reference to it (via a `wgpu::Buffer`, say).
+///
+/// - Other resources may depend on it (a texture view's backing
+///   texture, for example).
+///
+/// - It may be used by commands sent to the GPU that have not yet
+///   finished execution.
+///
+/// [`Device`]: crate::device::resource::Device
+/// [`Buffer`]: crate::resource::Buffer
+#[derive(Debug)]
+pub struct ResourceInfo<Id: TypedId> {
+    id: RwLock<Option<Valid<Id>>>,
```

Comment:
This `RwLock` doesn't seem necessary. The only places that ever call `set_id` are `Storage::insert` and `Storage::force_replace`, and they both have ownership of the value at that point.

Instead, this field should just be `Option<Valid<Id>>`, `Resource` should have an `as_info_mut` method that returns an `&mut ResourceInfo`, and a `set_id` method with a default definition that just calls `as_info_mut` and stores the new id.